### DISCUSSION TITLE
CPI Radiocontrols data model

### DISCUSSION THREAD

**Discussion by jarufe** [oreplay/oreplay#3] [Ideas]
CPI (Control de Puntos Intermedios, Intermediate Control Points) is a radiocontrol system. A Radiocontrol is an application running on an Android smartphone. The data is stored and processed on a Java application server with a MySQL database. The web client is also based on Java, using JSP and Servlets.

The reason for opening this discussion is to show the data model used by the system, in case it could serve as inspiration for the **OReplay** data model. As a place to host data from radio controls, it would only need the `Event` (**evento**), `FollowUp` (**seguimiento**) and `User` (**usuario**) entities. The `event` is the main pillar. A `user` creates an `event` to be able to store `punches`, that is, readings sent by the radiocontrols. 

The only data necessary for tracking are: `station`, `SiCard` and `time of passage`. The entity is denormalized to be able to contain some additional data that can be used directly: the `runner` (**corredor**) it corresponds to, `team` (if it exists), `club`, `category` or `relay leg`. 

In this scheme, the `user` only exists associated with one `event`. This means that the existence of a `user` who is the owner of several `events` is not considered. Even if the person is the same, every time an `event` is created a `user` will be associated with it.

The CPI system allows for integral event management. This means that data can be stored from: `Categories`, `Clubs`, `Teams`, `Runners` (corredor), Control Points with follow up, `Types of control points` (**tipo de punto**) (check, clear, start, finish, normal) and the association of `Control Points` (**punto de control**) to `Categories` (**categorías**), whether for classic events (Foot-O, MTBO) or for relay events.

There are other tables that were created to be part of a larger event monitoring system and that allow the calculations of results from classic, relay and rogaine events to be stored, but in reality the calculations can be carried out in real time and do not require these tables.

For OReplay, some of the tables will be conceptually common, but the schema will probably be more complex and complete.

![Modelo CPI](https://github.com/oreplay/oreplay/assets/17064412/2dc1172a-8550-4181-8332-ea97a4c4ba8b)

**Comment by freefri** [comment]
@jarufe Why do you need a different relation table CPs-Category for relays and for "normal" races?

What is RespuestasANT entity for?

Are punches stored as followUp entities?

I am trying to make a diagram with a bit more of info that the one you provided https://docs.google.com/drawings/d/134r2aN6ff1kas12PJIpl8G0VlPIEyt-4MnsgYgsppOA/edit

![image](https://github.com/oreplay/oreplay/assets/2754631/98775ff3-1acb-4b18-8a6e-5ef42b2a30c9)

**Reply by jarufe** [reply]
1.- Different relations for normal and relay. Each of them has some fields to describe the relationship and their are different. The main difference is that the relationship for relays has the leg number. It's perfectly possible to mix them in only one. In fact, if we keep the leg number, we could reuse the same control for several passings in normal races (now, this is not supported, and, now, you can not use a control that is visited more than once in a course by a runner)
2.- RespuestasANT stores the data coming from ANT app (Trail-O). If you use ANT (http://ant.yq.cz/) to manage timed groups (also to manage the answers for normal controls) you can send the data to CPI server. ControlOPrecision can download those data and compute results in real time, for example to show intermediate results as they are happening.
3.- FollowUp. Yes, it stores the punches coming from the radiocontrols. The basic data are: station, sicard and time. If you manage a more complex event, having the classes, runners, clubs, teams, etc, this related data is also stored in each record.
